module.c

```diff
@@ -119,98 +117,61 @@ static inline int mod_cmp_ext(const char *file, const size_t len,
   return len >= ext_len && !strcmp(file + len - ext_len, ext);
 }
 
-int mod_has_module_ext(const char *file, const size_t len, size_t *ext_pos)
+/*
+ * Return length of module file name without module extension if it is a
+ * module file name. Otherwise 0.
+ */
+int mod_basename_len(const char *file)
 {
   const mod_extensions_t *mod_ext;
+  size_t len = strlen(file);
 
-  for (mod_ext = mod_extensions; mod_ext->ext; mod_ext++)
-    if (mod_cmp_ext(file, len, mod_ext->ext, mod_ext->len)) {
-      if (ext_pos)
-	*ext_pos = len - mod_ext->len;
-      return 1;
+  for(mod_ext = mod_extensions; mod_ext->ext; mod_ext++) {
+    if(mod_cmp_ext(file, len, mod_ext->ext, mod_ext->len)) {
+      return len - mod_ext->len;
     }
+  }
 
   return 0;
 }
 
-int mod_find_module(const char *prefix, const char *module, char *file)
+
+/*
+ * Return file name with module extension removed if it is a module; else
+ * return a copy of the file name.
+ *
+ * The returned storage must be released using free().
+ */
+char *mod_basename(const char *file)
 {
-  const mod_extensions_t *mod_ext;
-  int file_len;
+  char *buf = strdup(file);
+  int base_len = mod_basename_len(file);
 
-  file_len = sprintf(file, "%s/%s", prefix, module);
-  for (mod_ext = mod_extensions; mod_ext->ext; mod_ext++) {
-    strcpy(file + file_len, mod_ext->ext);
-    if(util_check_exist(file))
-      return 1;
-  }
+  if(base_len) buf[base_len] = 0;
```

Comment:
If strdup() failed (out of memory), this will segfault.
I suggest to check buf immediately after strdup() and return 0 if buf == 0.